// ng2-bootstrap/components/buttons/button-radio.js
System.register(['angular2/angular2'], function(exports_1) {
    var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
        var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
        if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
        else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
        return c > 3 && r && Object.defineProperty(target, key, r), r;
    };
    var __metadata = (this && this.__metadata) || function (k, v) {
        if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
    };
    var __param = (this && this.__param) || function (paramIndex, decorator) {
        return function (target, key) { decorator(target, key, paramIndex); }
    };
    var angular2_1;
    var ButtonRadio;
    return {
        setters:[
            function (angular2_1_1) {
                angular2_1 = angular2_1_1;
            }],
        execute: function() {
            ButtonRadio = (function () {
                function ButtonRadio(cd, el) {
                    this.cd = cd;
                    this.el = el;
                    this.onChange = function (_) { };
                    this.onTouched = function () { };
                    // hack!
                    cd.valueAccessor = this;
                }
                ButtonRadio.prototype.onInit = function () {
                    this.uncheckable = typeof this.uncheckable !== 'undefined';
                };
                Object.defineProperty(ButtonRadio.prototype, "isActive", {
                    get: function () {
                        return this.btnRadio === this.value;
                    },
                    enumerable: true,
                    configurable: true
                });
                Object.defineProperty(ButtonRadio.prototype, "value", {
                    // hack view model!
                    get: function () {
                        return this.cd.viewModel;
                    },
                    set: function (value) {
                        this.cd.viewModel = value;
                        // hack: host classes updated before value is set >.<
                        if (this.isActive) {
                            this.el.nativeElement.classList.add('active');
                        }
                        else {
                            this.el.nativeElement.classList.remove('active');
                        }
                    },
                    enumerable: true,
                    configurable: true
                });
                // view -> model
                ButtonRadio.prototype.onClick = function () {
                    if (this.uncheckable && this.btnRadio === this.value) {
                        return this.cd.viewToModelUpdate(null);
                    }
                    this.cd.viewToModelUpdate(this.btnRadio);
                };
                // ControlValueAccessor
                // model -> view
                ButtonRadio.prototype.writeValue = function (value) {
                    this.value = value;
                };
                ButtonRadio.prototype.registerOnChange = function (fn) {
                    this.onChange = fn;
                };
                ButtonRadio.prototype.registerOnTouched = function (fn) {
                    this.onTouched = fn;
                };
                ButtonRadio = __decorate([
                    angular2_1.Directive({
                        selector: '[btn-radio][ng-model]',
                        properties: ['btnRadio', 'uncheckable'],
                        host: {
                            '(click)': 'onClick()',
                            '[class.active]': 'isActive'
                        }
                    }),
                    __param(0, angular2_1.Self()), 
                    __metadata('design:paramtypes', [(typeof (_a = typeof angular2_1.NgModel !== 'undefined' && angular2_1.NgModel) === 'function' && _a) || Object, (typeof (_b = typeof angular2_1.ElementRef !== 'undefined' && angular2_1.ElementRef) === 'function' && _b) || Object])
                ], ButtonRadio);
                return ButtonRadio;
                var _a, _b;
            })();
            exports_1("ButtonRadio", ButtonRadio);
        }
    }
});
